### PR TITLE
 Ray’s shared memory object store not garbage collecting after job completes

### PR DESCRIPTION
**Pull request by pfaraone**
> When a worker calls `ray.put()` or returns from a task, it copies the provided values into Ray’s shared memory object store. Ray will then [make these objects available](https://docs.google.com/document/d/1tBw9A4j62ruI5omIJbMxly-la5w4q_TjyJgJL_jN2fI/preview#heading=h.rtepjltxoeb7) throughout the cluster, attempt to [recover](https://docs.google.com/document/d/1tBw9A4j62ruI5omIJbMxly-la5w4q_TjyJgJL_jN2fI/preview#heading=h.tlm9n2vnrd30) them in case of a failure, [spill](https://docs.google.com/document/d/1tBw9A4j62ruI5omIJbMxly-la5w4q_TjyJgJL_jN2fI/preview#heading=h.7ce0zjt0shwh) them if the object store exceeds its configured capacity, and [garbage-collect](https://docs.google.com/document/d/1tBw9A4j62ruI5omIJbMxly-la5w4q_TjyJgJL_jN2fI/preview#heading=h.pyvgsp25rn5q) them once all ObjectRefs have gone out of scope.

https://docs.ray.io/en/latest/ray-core/scheduling/memory-management.html#objectref-reference-counting

Ray’s object store ([Plasma Object Store](https://docs.ray.io/en/latest/ray-core/objects/serialization.html#plasma-store)) only contains data stored via ray.put(), task args and return values. Ray implements distributed reference counting so that any ObjectID in scope in the cluster is pinned in the object store. This includes local python references, arguments to pending tasks, and IDs serialized inside of other objects. However, even after the job completes, the plasma store usage remains elevated as indicated when running `ray memory` in the head node.

In this PR, these changes result in the object emptying after the job completes: 
1. Manually deleting the cloudpickled object references that are pinned in memory. Calling `del` on a Python object decreases the local reference count by 1.

These changes seem to fix Ray distributed reference counting of the object references (allowing it to hit 0) and thereby allowing the object to be garbage collection. 

Additionally, imports in files touched were sorted using `isort`

### CODE REVIEW

File: deltacat/compute/compactor/steps/dedupe.py [hunk 1]
```diff
@@ -363,8 +362,9 @@ def dedupe(
         src_file_records_obj_refs.append(object_ref)
         mat_bucket_to_dd_idx_obj_id[mat_bucket] = (
             dedupe_task_index,
-            cloudpickle.dumps(object_ref),
+            object_ref,
         )
+        del ray.put
```
Comment (raghumdani) [inline deltacat/compute/compactor/steps/dedupe.py:367]: del ray.put?

File: deltacat/compute/compactor/utils/primary_key_index.py [hunk 2]
```diff
@@ -209,7 +208,8 @@ def group_hash_bucket_indices(
         if obj is not None:
             obj_ref = ray.put(obj)
             object_refs.append(obj_ref)
-            hash_bucket_group_to_obj_id[hb_group] = cloudpickle.dumps(obj_ref)
+            hash_bucket_group_to_obj_id[hb_group] = obj_ref
+            del obj_ref
```
Comment (raghumdani) [inline deltacat/compute/compactor/utils/primary_key_index.py:212]: Though this approach is cleaning up object store, its not cleaning up main memory.

File: deltacat/compute/compactor/steps/dedupe.py (hunk 1 above)
Comment (pfaraone) [reply deltacat/compute/compactor/steps/dedupe.py:367]: Type - fixed now

File: deltacat/compute/compactor/utils/primary_key_index.py (hunk 2 above)
Comment (pfaraone) [reply deltacat/compute/compactor/utils/primary_key_index.py:212]: I've experimented with the free internal API but it's not having any impact on cleaning up main memory.

**Comment by valiantljk** [comment]
just reviewed the changes. overall, looks good. One question, are there any other places in the code that use cloudpickle? should we change all?

**Comment by pfaraone** [comment]
> just reviewed the changes. overall, looks good. One question, are there any other places in the code that use cloudpickle? should we change all?

In this PR, all `cloudpickle` API calls have been removed.

File: deltacat/compute/compactor/utils/primary_key_index.py (hunk 2 above)
Comment (pfaraone) [reply deltacat/compute/compactor/utils/primary_key_index.py:212]: Running  `ray globalgc` has no impact on the main memory usage as well.

**Review by pdames — COMMENTED** [review]
Looks OK at a high-level, but a couple overall comments/questions:

1. Have we run benchmarks with the new code to assess its performance, memory consumption, runtime latency, stability etc. vs. the old? What were the results?

2. I feel like, if we really want to ensure a clean state at the beginning of the next round, our efforts would be better spent in launching a new cluster per round, and then focusing more on removing unnecessary memory consumed between each high-level parallel step (hash bucket, dedupe, materialize). Even if we make some incremental improvements to object store memory consumption between rounds, we're still left with a dirty cluster at the start of the next round. The persistence of anything written to disk, metadata written to the GCS, or any other untracked artifacts on any of our cluster nodes seem to make per-round cluster cleanup a moving target.

File: deltacat/compute/compactor/steps/dedupe.py [hunk 3]
```diff
@@ -363,8 +362,9 @@ def dedupe(
         src_file_records_obj_refs.append(object_ref)
         mat_bucket_to_dd_idx_obj_id[mat_bucket] = (
             dedupe_task_index,
-            cloudpickle.dumps(object_ref),
+            object_ref,
         )
+        del object_ref
```
Comment (pdames) [inline deltacat/compute/compactor/steps/dedupe.py:367]: This and our other manual object ref deletes will look confusing to future maintainers who don't have all the context that preceded this PR. Can you include a comment stating why we're including the delete here, what it is meant to help with, and/or a link to an existing discussion about the same on GitHub or the Ray Discussion forums?

File: deltacat/compute/compactor/utils/primary_key_index.py (hunk 2 above)
Comment (raghumdani) [reply deltacat/compute/compactor/utils/primary_key_index.py:212]: Makes sense. Let's revisit if it becomes a blocker.

File: deltacat/compute/compactor/steps/dedupe.py (hunk 3 above)
Comment (pfaraone) [reply deltacat/compute/compactor/steps/dedupe.py]: Resolving this conversation as I'm no longer using the `del` for manual object deletion in my latest commit. Solely removing the `cloudpickle` API calls allows the object memory to free up properly after the job run finishes.

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -11,8 +11,6 @@ repos:
     -   id: end-of-file-fixer
     -   id: requirements-txt-fixer
     -   id: trailing-whitespace
-    -   id: no-commit-to-branch
```
Comment (pfaraone) [inline .pre-commit-config.yaml:14]: There are alternative ways to protect specific branches from direct checkins using only Github settings.

**Review by raghumdani — APPROVED** [review]
LGTM

File: deltacat/compute/compactor/utils/primary_key_index.py [hunk 5]
```diff
@@ -206,11 +206,13 @@ def group_hash_bucket_indices(
...
-            object_refs.append(obj_ref)
-            hash_bucket_group_to_obj_id[hb_group] = cloudpickle.dumps(obj_ref)
-
+        if obj is None:
+            continue
+        obj_ref = ray.put(obj)
+        pickled_obj_ref = cloudpickle.dumps(obj_ref)
+        object_refs.append(pickled_obj_ref)
+        hash_bucket_group_to_obj_id[hb_group] = pickled_obj_ref
+        del obj_ref
```
Comment (raghumdani) [inline deltacat/compute/compactor/utils/primary_key_index.py:215]: Let's add documentation around why we are deleting this object reference.

File: deltacat/compute/compactor/steps/materialize.py [hunk 6]
```diff
@@ -90,9 +92,9 @@ def _materialize(
     logger.info(f"Starting materialize task...")
     dedupe_task_idx_and_obj_ref_tuples = [
         (
-            t[0],
-            cloudpickle.loads(t[1])
-        ) for t in dedupe_task_idx_and_obj_id_tuples
```
Comment (pfaraone) [inline deltacat/compute/compactor/steps/materialize.py:95]: Refactor only

**Review by rkenmi — APPROVED** [review]

**Pull request merged**